soccer/gameplay/situational_play_selection.py

```diff
@@ -298,60 +318,64 @@ def isMidfield(self):
 
-    ##Update determining if we want to preempt the current play or not
-    #
-    # Preemption is still an open question but this is a prototype of
-    # of how a non-invasive preemption system might work
-    #
-    def updatePreempt(self):
-        if (self.lastSituation != self.currentSituaion and
-                self.situationChangeTime == None):
-            self.situationChangeTime = time.time()
+    ##Returns true if we are in any of the situations in the passed list
+    #passed situation list must be enums
+    def isSituations(self, situations):
+        return self.currentSituation in situations
 
-        if (self.currentPreempt):
-            self.currentPreempt = False
-        elif (self.situationChangeTime != None):
-            if (abs(time.time() - self.situationChangeTime) >
-                    self.playPreemptTime):
-                self.situationChangeTime = None
-                self.currentPreempt = True
-                self.LastSituation = self.currentSituaion
+    ##
+    # 
+    # Will determine if the current play has lasted too long over a situation change
+    #   and needs to be preempted, it will call try_preempt
+    def updatePreempt(self):
 
-    '''def addPreempt(play) possibly a function to add transition out of every state to the completed state with preemptPlay as the lambda
-        for g in states:
-            play.add_transition(g -> completed , preemptPlay)'''
+        #Get the current play, may want to add a getter, as this is a "private" variable
+        currentPlay = main._root_play.play
 
-    ##A function to determine if the currently running play should be preempted
-    def preemptPlay(self):
-        return self.currentPreempt
+        #If the last situation is not the current situation, and the situationChanged flag
+        #is not already up, put it up and record the time, and change last situation to be the current situation
+        if (self.lastSituation != self.currentSituation and
+                not self.situationChanged):
+            self.situationChangeTime = time.time()
+            self.lastSituation = self.currentSituation
+            self.situationChanged = True
+
+        #If the play changes, put the situationChanged flag down and change lastPlay
+        if (self.lastPlay != currentPlay):
+            self.lastPlay = currentPlay
+            self.situationChanged = False
+
+        #If the situation has changed and its been longer than situationChange
+        #time without a play change, try to preempt the current play
+        if (self.situationChanged and
+                abs(time.time() - self.situationChangeTime) >
+                self.preemptTime):
+            #Check to see if the play is a standard play before trying to preempt
+            if (currentPlay != None and isinstance(
+                    currentPlay, standard_play.StandardPlay) and self.enabled):
+                if (currentPlay.try_preempt()):
+                    self.situationChanged = False
```

Comment:
This only asks the play if it should try to preempt once on situation change. Should we continuously do it? I imagine we might want to put off a preempt if a pass is happening or something, but will it be up to the play to then quit after that happens or should we just continue polling until that pass is finished

Reply:
So I had it preempting continuously up until a few days ago and I changed it to only do it a single time to as a quick fix to concerns I had about not detecting when a play is preempted and restarts, and continuously preempting it due to that (because we only check root play equality, which will still be true after switching). I plan to switch back to continuously preempting in the next iteration alongside a more robust detection for changes. This is just a beta that is disabled by default.